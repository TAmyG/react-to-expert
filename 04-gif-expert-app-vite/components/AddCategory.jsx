import { useState } from "react";
import PropTypes from 'prop-types';


/**
 * 
 * Mostrar error para mostrar el onChange
 * 1. Crear state
 * 2. agregar input con su onInputChange
 * 3. Agregar form
 * 4. Creare onSubmit
 * 5. Agregar prop setCategories
 * 6. Tarea agregar inputValue a setCategories y definir proptype
 * 7. Implementar onNewCategory
 */
export const AddCategory = ({setCategories,onNewCategory}) => {

    const [inputValue, setInputValue] = useState('');

    const onInputChange = (e) => {
        //setInputValue('Hola mundo');
        //console.log(e.target.value);
        setInputValue(e.target.value);
    }

    const onSubmit = (e)=>{
        e.preventDefault();
        if(inputValue.trim().length <= 1) return;

        //1.React no redenriza por cada ejecucion del state sino al final de que la funcion principal termine todas sus operaciones
        //setCategories(categories=> [inputValue, ...categories]);
        //2.
        onNewCategory(inputValue.trim());


        setInputValue('');
    }


    return (
        <form onSubmit={onSubmit}>
            <input
                type="text"
                placeholder="Buscar Gifs"
                value={inputValue}
                // Explicar atomicamente asi se pasa la referencia de una funcion 
                onChange={(event) => onInputChange(event)}
            />
        </form>
    )
}


AddCategory.propTypes = {
    //setCategories: PropTypes.func.isRequired
}