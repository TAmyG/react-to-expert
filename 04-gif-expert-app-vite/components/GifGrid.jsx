import { useState } from "react";
import { useEffect } from "react";
import { getGifs } from "../helpers/getGifs";

/**
 * 1. ejemplo arreglo de gifs
 * 2. implementacion de API giphy con getGifs
 * 3. implementacion useEffect
 * 4. listar titulos de gifs
 * @param {category} param0 
 * @returns 
 */
export const GifGrid = ({ category }) => {

    //1.
    //const gifs = [1,2,3,4,5];
    //2.
    // getGifs();//Nunca colocar llamadas directas en el componente por el renderizado

    //3.
    const [gifs, setGifs] = useState([]);

    const getLocalGifs = async () => {
        const newGifs = await getGifs(category);
        setGifs(newGifs);
    }

    useEffect(() => {
        getLocalGifs();
    }, [gifs])

    return (
        <>
            <h3>{category}</h3>
            {/* {gifs.map(gif=><p key={gif}>{gif}</p>)} */}
            <ol>
                {
                    gifs.map(gif => <li key={gif.id}>{gif.title}</li>)
                }
            </ol>


        </>
    )
}
