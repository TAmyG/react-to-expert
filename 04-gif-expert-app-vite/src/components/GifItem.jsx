import PropTypes from 'prop-types';
/**
 * 1. colocar <li>
 * 2. crear componente
 * @param {title, id, url} param0 
 * @returns 
 */
export const GifItem = ({title, id, url}) => {
  return (
    // <li >{title}</li>
    <div className='card'>
        <img src={url} alt={title}/>
        <p>{title}</p>
    </div>
  )
}


GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  id: PropTypes.string,
}