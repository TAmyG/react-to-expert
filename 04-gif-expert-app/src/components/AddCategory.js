import React, { useState } from 'react';
import PropTypes from 'prop-types';

// Destructure of props
export const AddCategory = ({ setCategories }) => {
    const [inputValue, setInputValue] = useState('');

    const handleInputChande = (e) => {
        console.log(e.target.value);
        setInputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setCategories((cat) => [inputValue, ...cat]);
            setInputValue('');
        }
    };

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChande}
            />
        </form>
    );
};

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
};
