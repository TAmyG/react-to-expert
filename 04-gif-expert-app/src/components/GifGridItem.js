import React from 'react';
import PropTypes from 'prop-types';

export const GifGridItem = ({ id, title, url }) => {
    console.log(id, title, url);
    return (
        <div className="card animate__animated animate__fadeIn animate__faster">
            <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    );
};

GifGridItem.propTypes = {
    url: PropTypes.string,
    title: PropTypes.string,
};
