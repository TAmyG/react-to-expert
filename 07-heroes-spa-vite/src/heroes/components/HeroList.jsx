import { useMemo } from "react";
import { getHeroesByPublisher } from "../helpers";
import { HeroCard } from "./HeroCard";

/**
 * 
 * Tarea que implementen el herolist
 * Implementar card hero
 * Implementar use memo
 */
export const HeroList = ({publisher}) => {

    const heroes = useMemo(()=> getHeroesByPublisher(publisher), [publisher]);
  return (
    <div className="row rows-cols-1 row-cols-md-3 g-3">
        {
            heroes.map(hero=> 
                // <li key={hero.id}>{hero.superhero}</li>
                <HeroCard key={hero.id} {...hero}/>
            )
        }
    </div>
  )
}
