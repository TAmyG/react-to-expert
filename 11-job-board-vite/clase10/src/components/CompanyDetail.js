import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import { JobList } from "./job/JobList";
import { useDispatch, useSelector } from "react-redux";
import { acLoadCompany } from "../actions/companyAction";

export const CompanyDetail = () => {
  const { companyId } = useParams();

  const dispatch = useDispatch();
  const { company } = useSelector((state) => state.company);

  useEffect(() => {
    dispatch(acLoadCompany(companyId));
  }, [companyId, dispatch]);

  return (
    <>
      {company && (
        <div>
          <h1 className="title">{company.name}</h1>
          <div className="box">{company.description}</div>

          <h1 className="title is-5">Jobs at {company.name}</h1>
          <JobList jobs={company.jobs} />
        </div>
      )}
    </>
  );
};
