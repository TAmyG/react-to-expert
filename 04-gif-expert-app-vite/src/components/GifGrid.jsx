import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifItem } from "./GifItem";

/**
 * 1. ejemplo arreglo de gifs
 * 2. implementacion de API giphy con getGifs
 * 3. implementacion useEffect
 * 4. listar titulos de gifs
 * 5. crear componente GifItem
 * 6. crear custom hook useFetchGifs
 * @param {category} param0 
 * @returns 
 */
export const GifGrid = ({ category }) => {

    //1.
    //const gifs = [1,2,3,4,5];
    //2.
    // getGifs();//Nunca colocar llamadas directas en el componente por el renderizado

    //3.
    // const [gifs, setGifs] = useState([]);

    // const getLocalGifs = async () => {
    //     const newGifs = await getGifs(category);
    //     //console.log(newGifs); 
    //     setGifs(newGifs);
    // }

    // useEffect(() => {
    //     getLocalGifs();
    // }, []);
    //////////////////////////////////////////////
    //6.
    const {gifs, isLoading} = useFetchGifs(category);

    return (
        <>
            <h3>{category}</h3>
            {/* {gifs.map(gif=><p key={gif}>{gif}</p>)} */}
            {/* <ol> */}
            {
                isLoading && <h1>Loading...</h1>
            }
            <div className="card-grid">
            
                {
                    gifs.map(gif => <GifItem key={gif.id} {...gif}/>)
                }
            </div>
            {/* </ol> */}
        </>
    )
}
