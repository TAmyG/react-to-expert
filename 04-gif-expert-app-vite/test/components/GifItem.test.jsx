import { screen, render } from '@testing-library/react';
import { GifItem } from '../../src/components/GifItem';

describe('Pruebas en <GifItem />', () => {
    const title = 'Hello world';
    const url = 'https://media1.giphy.com/media/2y98KScHKeaQM/giphy.gif?cid=39ae98e5q29hy9kunadf2qo0a12204ne9e8hkjhfah9bago1&rid=giphy.gif&ct=g';

    test('debe hacer match en el snapshot', () => {
        const { container } = render(<GifItem title={title} url={url} />);
        expect(container).toMatchSnapshot();
    });

    test('debe mostrar la imagen con el URL y el ALT', () => {
        render(<GifItem title={title} url={url} />);
        // expect(screen.getByRole('img').src).toContain(url);
        const {src, alt}= screen.getByRole('img');
        expect(src).toBe(url);
        expect(alt).toBe(title);
    });


    test('debe mostrar el titulo en el componente', () => {
        render(<GifItem title={title} url={url} />);
        
        expect(screen.getByText(title)).toBeTruthy();
    });
})