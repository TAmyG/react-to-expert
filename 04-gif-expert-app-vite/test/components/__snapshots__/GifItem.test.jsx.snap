// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem /> debe hacer match en el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Hello world"
      src="https://media1.giphy.com/media/2y98KScHKeaQM/giphy.gif?cid=39ae98e5q29hy9kunadf2qo0a12204ne9e8hkjhfah9bago1&rid=giphy.gif&ct=g"
    />
    <p>
      Hello world
    </p>
  </div>
</div>
`;
